src/mds/Beacon.cc

```diff
@@ -514,9 +514,19 @@ void Beacon::notify_health(MDSRank const *mds)
         MDSHealthMetric m(MDS_HEALTH_CLIENTS_LAGGY, HEALTH_WARN, css->strv());
         laggy_clients_metrics.emplace_back(std::move(m));
       }
-      auto&& m = laggy_clients_metrics;
-      health.metrics.insert(std::end(health.metrics), std::cbegin(m),
-                            std::cend(m));
+
+      if (laggy_clients.size() <= (size_t)g_conf()->mds_health_summarize_threshold) {
+	auto&& m = laggy_clients_metrics;
+	health.metrics.insert(std::end(health.metrics), std::cbegin(m),
+			      std::cend(m));
+      } else {
+	CachedStackStringStream css;
+	*css << "Many client (" << laggy_clients.size()
+	     << ") are laggy; not evicted since some OSD(s) are laggy";
+	MDSHealthMetric m(MDS_HEALTH_CLIENTS_LAGGY_MANY, HEALTH_WARN, css->strv());
+	m.metadata["client_count"] = stringify(laggy_clients.size());
+	health.metrics.push_back(std::move(m));
+      }
```

Comment:
so this is inside the `if (defer_client_eviction && !laggy_clients.empty()) {`, and there will be this for loop executed:
```
      for (const auto& laggy_client: laggy_clients) {
        CachedStackStringStream css;
        *css << "Client " << laggy_client << " is laggy; not evicted"
             << " because some OSD(s) is/are laggy";
        MDSHealthMetric m(MDS_HEALTH_CLIENTS_LAGGY, HEALTH_WARN, css->strv());
        laggy_clients_metrics.emplace_back(std::move(m));
      }
```

and this code blocks compares the num of clients laggy against the threshold `mds_health_summarize_threshold`, if the count is more then it also would add `MDS_HEALTH_CLIENTS_LAGGY_MANY` to the cluster health, isn't this a bit off-putting i.e. you might send `MDS_HEALTH_CLIENTS_LAGGY` and `MDS_HEALTH_CLIENTS_LAGGY_MANY` both? instead this check can be shifted prior to the loop i mentioned, what do you think?

Reply:
I don't think that's the case. `health.metrics` is updated only in the `if` condition that's added. Not sure why you say it's off-putting o_O

Reply:
@vshankar this will be the final block for laggy clients
```
    if (defer_client_eviction && !laggy_clients.empty()) {
      std::vector<MDSHealthMetric> laggy_clients_metrics;
      for (const auto& laggy_client: laggy_clients) {
        CachedStackStringStream css;
        *css << "Client " << laggy_client << " is laggy; not evicted"
             << " because some OSD(s) is/are laggy";
        MDSHealthMetric m(MDS_HEALTH_CLIENTS_LAGGY, HEALTH_WARN, css->strv());
        laggy_clients_metrics.emplace_back(std::move(m));
      }
      if (laggy_clients.size() <= (size_t)g_conf()->mds_health_summarize_threshold) {
	      auto&& m = laggy_clients_metrics;
	      health.metrics.insert(std::end(health.metrics), std::cbegin(m),
			  std::cend(m));
      } else {
	      CachedStackStringStream css;
	      *css << "Many client (" << laggy_clients.size()
	      << ") are laggy; not evicted since some OSD(s) are laggy";
	      MDSHealthMetric m(MDS_HEALTH_CLIENTS_LAGGY_MANY, HEALTH_WARN, css->strv());
	      m.metadata["client_count"] = stringify(laggy_clients.size());
	      health.metrics.push_back(std::move(m));
      }
    }
```

for every laggy we add a `MDS_HEALTH_CLIENTS_LAGGY`, and then if the num of laggy clients exceeds the threshold then we also add `MDS_HEALTH_CLIENTS_LAGGY_MANY`. So having two warnings for laggy clients. Instead what I meant is let us first check for the threshold and depending on that either we say "many clients laggy" or iterate over all and say "client [1-n] laggy"

Reply:
That's added in `laggy_clients_metrics` which pushed in `health.metrics` if the size of `laggy_clients_metrics` exceeds `mds_health_summarize_threshold`.

Reply:
@dparmar18 `MDSHealthMetric m` is local to the loop where `laggy_clients` is iterated. Maybe I'm not understanding your point...

Reply:
AH, I understand your point now. Makes sense.